src/core/gltf2importer/meshparser.cpp

```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+        const Qt3DRender::QAttribute *uvAttribute = nullptr;
+        Qt3DRender::QAttribute *tangentAttribute = nullptr;
+    };
+
+    MikkTSpaceUserData userData;
+    const auto &attributes = geometry->attributes();
+    std::for_each(std::begin(attributes), std::end(attributes), [&userData] (const Qt3DRender::QAttribute *attr) {
+        if (attr->attributeType() == Qt3DRender::QAttribute::IndexAttribute)
+            userData.indexAttribute = attr;
+        if (attr->name() == Qt3DRender::QAttribute::defaultPositionAttributeName())
+            userData.positionAttribute = attr;
+        if (attr->name() == Qt3DRender::QAttribute::defaultNormalAttributeName())
+            userData.normalAttribute = attr;
+        if (attr->name() == Qt3DRender::QAttribute::defaultTextureCoordinateAttributeName())
+            userData.uvAttribute = attr;
+    });
+
+    if (!userData.indexAttribute ||
+        !userData.positionAttribute ||
+        !userData.normalAttribute ||
+        !userData.uvAttribute)
+        return nullptr;
+
+    // Create the tangent attribute
+    QByteArray tangentBufferData;
+    tangentBufferData.resize(userData.positionAttribute->count()*sizeof(QVector4D));
+    auto *tangentBuffer = new Qt3DRender::QBuffer;
+    tangentBuffer->setData(tangentBufferData);
+    auto *tangentAttribute = new Qt3DRender::QAttribute(tangentBuffer,
+                                                        Qt3DRender::QAttribute::Float,
+                                                        4,
+                                                        userData.positionAttribute->count());
+    userData.tangentAttribute = tangentAttribute;
+
+    //Create callbacks for MikkTSpace
+    SMikkTSpaceContext mikkContext;
+    mikkContext.m_pUserData = &userData;
+
+    SMikkTSpaceInterface interface;
+    interface.m_getNumFaces = [](const SMikkTSpaceContext *pContext) -> int {
+        const auto *userData = reinterpret_cast<MikkTSpaceUserData*>(pContext->m_pUserData);
+        return userData->indexAttribute->count() / 3;
+    };
+
+    interface.m_getNumVerticesOfFace = [](const SMikkTSpaceContext *pContext, const int iFace) ->int {
+        return 3;
+    };
+
+    interface.m_getPosition = [] (const SMikkTSpaceContext *pContext,
+                    float fvPosOut[],
+                    const int iFace,
+                    const int iVertex) {
+        const auto *userData = reinterpret_cast<MikkTSpaceUserData*>(pContext->m_pUserData);
+        const auto vertexIndices = findVertexIndicesInFace(userData->indexAttribute, iFace);
+        const auto vertexIndex = vertexIndices[iVertex];
+        const auto vertexBuffer = userData->positionAttribute->buffer();
+        const auto byteOffset = userData->positionAttribute->byteOffset();
+        const auto byteStride = userData->positionAttribute->byteStride() == 0 ?
+                                3*4 : // Position vertex base type must be float
```

Comment:
maybe an assert to ensure this ?

Reply:
Yes, but not here. Attributes have a specific type required by gltf. We should check there that we are using the correct types.

Reply:
I have added a new commit to the PR that checks if the mesh is valid. That way, when we generate the tangents, we have already checked all this.